### PR TITLE
Unify ESM failure destinations and Pipes

### PR DESCRIPTION
**Pull request by joe4dev**
Depends on: https://github.com/localstack/localstack/pull/11331

## Motivation

Lambda ESM v2 currently doesn't support failure destinations. This PR unifies Pipes dead letter config and Lambda ESM failure destinations. Unfortunately, there are some more subtle differences than I hoped for, but being able to convert an ESM config into a Pipe dead letter config in the factory is quite neat.

## Changes

* Add support for Lambda ESM failure destinations
* Unify `send_events_to_dlq` for ESM and Pipes and Kinesis and DynamoDB Streams
* Re-validate some snapshots & increase timeout for AWS (after running into flake)

## Testing

Using the existing Lambda ESM test suite to reach parity with ESM v1 for now.

## Notes

Context diff:

```py
# ESM has this extra
"requestContext": {
  "requestId": "<uuid:2>",
  "functionArn": "arn:<partition>:lambda:<region>:111111111111:function:<resource:3>",
  "condition": "RetryAttemptsExhausted",
  "approximateInvokeCount": 2
},
"responseContext": {
  "statusCode": 200,
  "executedVersion": "$LATEST",
  "functionError": "Unhandled"
},

# Pipes has this extra
"context": {
    "condition": "RetryAttemptsExhausted",
    "partnerResourceArn": "arn:<partition>:pipes:<region>:111111111111:pipe/<pipe-name>"
  },
```

## TODO

- [ ] Fix snapshot mismatches for `test_kinesis_event_source_mapping_with_on_failure_destination_config` and `test_dynamodb_event_source_mapping_with_on_failure_destination_config`
  - [ ] `ApproximateCreationDateTime` conversion in DynamoDB poller
  - [ ] Fix different context for Pipes and ESM
- [ ] Switch PR target to master once https://github.com/localstack/localstack/pull/11331 is merged

### CODE REVIEW

File: tests/aws/services/lambda_/event_source_mapping/test_lambda_integration_kinesis.py [hunk 1]
```diff
@@ -535,9 +546,9 @@ def verify_failure_received():
 # TODO: add tests for different edge cases in filtering (e.g. message isn't json => needs to be dropped)
 # https://docs.aws.amazon.com/lambda/latest/dg/invocation-eventfiltering.html#filtering-kinesis
 class TestKinesisEventFiltering:
-    @pytest.mark.skipif(
-        is_v2_esm(), reason="DestinationConfig and LastProcessing result snapshot diff in ESM v2"
-    )
+    # @pytest.mark.skipif(
+    #     is_v2_esm(), reason="DestinationConfig and LastProcessing result snapshot diff in ESM v2"
+    # )
```
Comment (dfangl) [inline tests/aws/services/lambda_/event_source_mapping/test_lambda_integration_kinesis.py:551]: Can this be deleted instead of commenting it out?

File: localstack-core/localstack/services/lambda_/event_source_mapping/pollers/stream_poller.py [hunk 2]
```diff
@@ -182,9 +194,12 @@ def poll_events_from_shard(self, shard_id: str, shard_iterator: str):
...
         # Send failed events to potential DLQ
         abort_condition = abort_condition or "RetryAttemptsExhausted"
+        # TODO: fix format for ESM using `requestContext` and `responseContext` (which requires info from the Lambda sender!) instead
         context = {
-            "condition": abort_condition,
-            "partnerResourceArn": self.partner_resource_arn,
+            "context": {
+                "condition": abort_condition,
+                "partnerResourceArn": self.partner_resource_arn,
+            }
```
Comment (dfangl) [inline localstack-core/localstack/services/lambda_/event_source_mapping/pollers/stream_poller.py:202]: So this will soon differ depending on ESM/pipes, right?

**Review by dfangl — APPROVED** [review]
LGTM, just some minor comments!

File: localstack-core/localstack/services/lambda_/event_source_mapping/esm_worker_factory.py [hunk 3]
```diff
@@ -98,13 +98,21 @@ def get_esm_worker(self) -> EsmWorker:
             )
         elif source_service == "kinesis":
             # TODO: map all supported ESM to Pipe parameters
+            optional_params = {}
+            dead_letter_config_arn = (
+                self.esm_config.get("DestinationConfig", {}).get("OnFailure", {}).get("Destination")
+            )
+            if dead_letter_config_arn:
+                optional_params["DeadLetterConfig"] = {"Arn": dead_letter_config_arn}
```
[the remark this answers is not in the document]
Comment (joe4dev) [reply localstack-core/localstack/services/lambda_/event_source_mapping/esm_worker_factory.py:106]: Nope, Lambda failure destinations map to Pipes DeadLetterConfig and success destinations are not yet implemented in ESM v2:

[CreateEventSourceMapping](https://docs.aws.amazon.com/lambda/latest/api/API_CreateEventSourceMapping.html#lambda-CreateEventSourceMapping-request-FunctionName)

```json
"DestinationConfig": { 
   "OnFailure": { 
      "Destination": "string"
   },
   "OnSuccess": { 
      "Destination": "string"
   }
},
```

[CreatePipe](https://docs.aws.amazon.com/eventbridge/latest/pipes-reference/API_CreatePipe.html)

```json
"DeadLetterConfig": { 
   "Arn": "string"
},
```

File: localstack-core/localstack/services/lambda_/event_source_mapping/pollers/dynamodb_poller.py [hunk 4]
```diff
@@ -84,14 +85,28 @@ def transform_into_events(self, records: list[dict], shard_id) -> list[dict]:
...
+
+    def failure_payload_details_field_name(self) -> str:
+        return "DDBStreamBatchInfo"
+
+    def get_approximate_arrival_time(self, record: dict) -> float:
+        # TODO: validate whether the default should be now
+        # Optional according to AWS docs:
+        # https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_streams_StreamRecord.html
+        # TODO: parse float properly if present from ApproximateCreationDateTime -> now works, compare via debug!
+        return record["dynamodb"].get("todo", datetime.utcnow().timestamp())
```
[the remark this answers is not in the document]
Comment (joe4dev) [reply localstack-core/localstack/services/lambda_/event_source_mapping/pollers/dynamodb_poller.py:109]: yes because using `ApproximateCreationDateTime` causes a parity issue and it's unclear whether that's the right field 🤔 .

Sidenote: ESM v1 attempts to get `ApproximateArrivalTimestamp` (which never exist for DynamoDB [GetRecords](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_streams_GetRecords.html)) and therefore always defaults to `datetime.datetime.utcnow()`

File: tests/aws/services/lambda_/event_source_mapping/test_lambda_integration_kinesis.py (hunk 1 above)
Comment (joe4dev) [reply tests/aws/services/lambda_/event_source_mapping/test_lambda_integration_kinesis.py:551]: ups, will look into it and either delete or keep it skipped 👀  (oversight in [62272ee](https://github.com/localstack/localstack/pull/11337/commits/62272eef1f3afe93e2119dd03c92f3d7579413e2))

File: localstack-core/localstack/services/lambda_/event_source_mapping/pollers/stream_poller.py (hunk 2 above)
Comment (joe4dev) [reply localstack-core/localstack/services/lambda_/event_source_mapping/pollers/stream_poller.py:202]: Correct, that's currently a parity issue in ESM v2. Difference:

```py
# ESM has this extra
"requestContext": {
  "requestId": "<uuid:2>",
  "functionArn": "arn:<partition>:lambda:<region>:111111111111:function:<resource:3>",
  "condition": "RetryAttemptsExhausted",
  "approximateInvokeCount": 2
},
"responseContext": {
  "statusCode": 200,
  "executedVersion": "$LATEST",
  "functionError": "Unhandled"
},

# Pipes has this extra
"context": {
    "condition": "RetryAttemptsExhausted",
    "partnerResourceArn": "arn:<partition>:pipes:<region>:111111111111:pipe/<pipe-name>"
  },
  ```

File: tests/aws/services/lambda_/event_source_mapping/test_lambda_integration_kinesis.py (hunk 1 above)
Comment (joe4dev) [reply tests/aws/services/lambda_/event_source_mapping/test_lambda_integration_kinesis.py:551]: I'll leave it skipped for now because it's failing for another reason (little progress, but JSON filtering is still an issue as mentioned).

**Review by dfangl — APPROVED** [review]
